### PR TITLE
[WIP] detect overlong/surrogate utf8

### PR DESCRIPTION
**Pull request by buixor**
Add a generic internal rule associated to detection of overlong/surrogate utf8

### CODE REVIEW

File: naxsi_src/naxsi_utils.c [hunk 1]
```diff
@@ -99,14 +99,88 @@ u_int naxsi_escape_nullbytes(ngx_str_t *str) {
...
+    else
+      return s;
+  }
+  return NULL;
+}
+
+
+/* unescape routine :
+ - returns number of nullbytes present 
+ - returns -1 if overlong utf8 sequence
```
Comment (jvoisin) [inline naxsi_src/naxsi_utils.c]: Isn't this comment wrong?

File: naxsi_src/naxsi_runtime.c [hunk 2]
```diff
@@ -1414,7 +1423,15 @@ ngx_http_basestr_ruleset_n(ngx_pool_t	*pool,
   NX_DEBUG(_debug_basestr_ruleset , NGX_LOG_DEBUG_HTTP, req->connection->log, 0, 
 	   "XX-checking %d rules ...", rules->nelts); 
 
-  
+  /* check for overlong/surrogate utf8 encoding */
+  if (ngx_utf8_check(name) != NULL) {
+    ngx_http_apply_rulematch_v_n(&nx_int__bad_utf8, ctx, req, NULL, NULL, zone, 1, 1);
+    return (0);
+  }
```
Comment (jvoisin) [inline naxsi_src/naxsi_runtime.c]: `else if` ?

File: naxsi_src/naxsi_utils.c [hunk 3]
```diff
@@ -99,14 +99,88 @@ u_int naxsi_escape_nullbytes(ngx_str_t *str) {
...
+	return s;
+      }
+      /* 1110XXXX 10Xxxxxx 10xxxxxx */
+      if ((s[1] & 0xc0) != 0x80 ||
+	  (s[2] & 0xc0) != 0x80 ||
+	  (s[0] == 0xe0 && (s[1] & 0xe0) == 0x80) ||    /* overlong? */
+	  (s[0] == 0xed && (s[1] & 0xe0) == 0xa0) ||    /* surrogate? */
+	  (s[0] == 0xef && s[1] == 0xbf &&
+	   (s[2] & 0xfe) == 0xbe))                      /* U+FFFE or U+FFFF? */
+	return s;
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c]: Add braces `{}` on that `else if`

File: naxsi_src/naxsi_utils.c [hunk 4]
```diff
@@ -99,14 +99,88 @@ u_int naxsi_escape_nullbytes(ngx_str_t *str) {
...
+    }
+    else if ((s[0] & 0xe0) == 0xc0) {
+      if (offset+1 >= str->len) {
+	//not enough bytes
+	return s;
+      }
+      /* 110XXXXx 10xxxxxx */
+      if ((s[1] & 0xc0) != 0x80 ||
+	  (s[0] & 0xfe) == 0xc0)                        /* overlong? */
+	return s;
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c]: Add braces `{}` on that `else if`

File: naxsi_src/naxsi_utils.c [hunk 5]
```diff
@@ -99,14 +99,88 @@ u_int naxsi_escape_nullbytes(ngx_str_t *str) {
...
+      }
+      /* 11110XXX 10XXxxxx 10xxxxxx 10xxxxxx */
+      if ((s[1] & 0xc0) != 0x80 ||
+	  (s[2] & 0xc0) != 0x80 ||
+	  (s[3] & 0xc0) != 0x80 ||
+	  (s[0] == 0xf0 && (s[1] & 0xf0) == 0x80) ||    /* overlong? */
+	  (s[0] == 0xf4 && s[1] > 0x8f) || s[0] > 0xf4) /* > U+10FFFF? */
+	return s;
+      else
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c]: Add braces `{}` on that `else`

File: naxsi_src/naxsi_utils.c [hunk 6]
```diff
@@ -99,14 +99,88 @@ u_int naxsi_escape_nullbytes(ngx_str_t *str) {
...
+	s += 4;
+    }
+    else
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c]: Add braces `{}` on that `else`

File: naxsi_src/naxsi_utils.c (hunk 3 above)
Comment (jvoisin) [reply naxsi_src/naxsi_utils.c]: We might want to keep the original formatting here, since it's "borrowed" code.

File: naxsi_src/naxsi_utils.c [hunk 7]
```diff
@@ -99,14 +99,88 @@ u_int naxsi_escape_nullbytes(ngx_str_t *str) {
...
+      return s;
+  }
+  return NULL;
+}
+
+
+/* 
+   unescape routine :
```
Comment (jvoisin) [inline naxsi_src/naxsi_utils.c]: How can you return two values in a `int` ;)

**Pull request merged**